gix-traverse/src/commit/mod.rs

```diff
@@ -0,0 +1,88 @@
+//! Provide multiple traversal implementations with different performance envelopes.
+//!
+//! Use [`Simple`] for fast walks that maintain minimal state, or [`Topo`] for a more elaborate traversal.
+use gix_hash::ObjectId;
+use gix_object::FindExt;
+use gix_revwalk::graph::IdMap;
+use gix_revwalk::PriorityQueue;
+use smallvec::SmallVec;
+
+/// A fast iterator over the ancestors of one or more starting commits.
+pub struct Simple<Find, Predicate> {
+    objects: Find,
+    cache: Option<gix_commitgraph::Graph>,
+    predicate: Predicate,
+    state: simple::State,
+    parents: Parents,
+    sorting: simple::Sorting,
+}
+
+/// Simple ancestors traversal, without the need to keep track of graph-state.
+pub mod simple;
+
+/// A commit walker that walks in topographical order, like `git rev-list
+/// --topo-order` or `--date-order` depending on the chosen [`topo::Sorting`].
+///
+/// Instantiate with [`topo::Builder`].
+pub struct Topo<Find, Predicate> {
+    commit_graph: Option<gix_commitgraph::Graph>,
+    find: Find,
+    predicate: Predicate,
+    indegrees: IdMap<i32>,
+    states: IdMap<topo::WalkFlags>,
```

Comment:
Do you think that `states` and `indegrees` could be merged? I am asking as these seem to be insert-only, and will use considerable amounts of memory on huge graphs.

Reply:
At first glance I don't see why not. But how much memory are we talking about? I'd guess the memory used by `IdMap<T1>` + `IdMap<T2>` wouldn't be that far off the amount used by `IdMap<(T1, T2)>`.

They are indeed insert-only. And that's just because in Git they are :stuck_out_tongue: There's a point where the indegree of a commit is set to zero. At that point I _think_ it can be removed from all the maps, but I'm not sure.

Reply:
As I mentioned in the discussion a couple of months ago (damn, time flies) I did try to add my stuff to `Ancestors` but couldn't get it to work well. I'm all for it, still, but I just don't see it. Doing it in `gix` seems fine to me.

Reply:
That's great to know (both :))!

This means if ever there is an issue with memory, there might be avenues to improve that. I assume Git also has multiple hashmaps for that? Or is it all in one there? * (if you recall, please ignore otherwise)*